### PR TITLE
IL-144: Add Unit Tests for Security packages: Api and Jwt

### PR DESCRIPTION
**Pull request by M437A**
Test for the security package
 - api package
 - jwt package

### CODE REVIEW

File: src/test/java/com/zufar/onlinestore/security/jwt/JwtTokenFromAuthHeaderExtractorTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class JwtTokenFromAuthHeaderExtractorTest {
+    @InjectMocks
+    private JwtTokenFromAuthHeaderExtractor jwtTokenFromAuthHeaderExtractor;
+    @Mock
+    private HttpServletRequest request;
+    private String jwtHttpRequestHeader = "Authorization";
```
Comment (plakhov) [inline src/test/java/com/zufar/onlinestore/security/jwt/JwtTokenFromAuthHeaderExtractorTest.java:27]: You can replace it through `HttpHeaders` class from Spring

**Review by andrew13pol — APPROVED** [review]

**Review by plakhov — APPROVED** [review]

File: src/test/java/com/zufar/onlinestore/security/api/CustomUserDetailsServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class CustomUserDetailsServiceTest {
+    @InjectMocks
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/CustomUserDetailsServiceTest.java:24]: add a space above

File: src/test/java/com/zufar/onlinestore/security/api/CustomUserDetailsServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    private CustomUserDetailsService customUserDetailsService;
+    @Mock
+    private UserRepository userRepository;
+    private String userEmail = Instancio.of(String.class)
+            .create();
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/CustomUserDetailsServiceTest.java:29]: please leave method on the same line 
`private String userEmail = Instancio.of(String.class).create();`

File: src/test/java/com/zufar/onlinestore/security/api/CustomUserDetailsServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    private UserEntity userDetails = Instancio.of(UserEntity.class)
+            .create();
+
+    @Test
+    @DisplayName("test load user by username")
+    void testLoadUserByUsername() {
+        when(userRepository.findByEmail(userEmail))
+                .thenReturn(Optional.ofNullable(userDetails));
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/CustomUserDetailsServiceTest.java:37]: please leave method one the same line 
` when(userRepository.findByEmail(userEmail)).thenReturn(Optional.ofNullable(userDetails));`

File: src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+class LoginAttemptFactoryTest {
+    private String userEmail = Instancio.of(String.class)
+            .create();
+
+    @Test
+    @DisplayName("test creating LoginAttemptEntity")
+    public void testLoginAttemptEntity(){
+        LoginAttemptEntity loginAttemptEntity =
+                LoginAttemptFactory.createInitialFailedLoggedAttemptEntity(userEmail);
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java:19]: leave on the same line

File: src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+        assertTrue(loginAttemptEntity!=null);
+        assertEquals(null, loginAttemptEntity.getId());
+        assertEquals(userEmail, loginAttemptEntity.getUserEmail());
+        assertEquals(0, loginAttemptEntity.getAttempts());
+        assertEquals(null, loginAttemptEntity.getExpirationDatetime());
+        assertEquals(false, loginAttemptEntity.getIsUserLocked());
+        assertTrue(loginAttemptEntity.getLastModified()!=null);
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java:27]: assertNotNull

File: src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    private String userEmail = Instancio.of(String.class)
+            .create();
+
+    @Test
+    @DisplayName("test creating LoginAttemptEntity")
+    public void testLoginAttemptEntity(){
+        LoginAttemptEntity loginAttemptEntity =
+                LoginAttemptFactory.createInitialFailedLoggedAttemptEntity(userEmail);
+
+        assertTrue(loginAttemptEntity!=null);
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java:21]: assertNotNull

File: src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        assertEquals(null, loginAttemptEntity.getId());
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java:22]: assertNull

File: src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        assertEquals(userEmail, loginAttemptEntity.getUserEmail());
+        assertEquals(0, loginAttemptEntity.getAttempts());
+        assertEquals(null, loginAttemptEntity.getExpirationDatetime());
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java:25]: assertNull

File: src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        assertEquals(false, loginAttemptEntity.getIsUserLocked());
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/LoginAttemptFactoryTest.java:26]: assertFalse

File: src/test/java/com/zufar/onlinestore/security/api/ResetLoginAttemptsServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.mockStatic;
+
+@ExtendWith(MockitoExtension.class)
+class ResetLoginAttemptsServiceTest {
+    @InjectMocks
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/ResetLoginAttemptsServiceTest.java:26]: add a space above

File: src/test/java/com/zufar/onlinestore/security/api/ResetLoginAttemptsServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    private ResetLoginAttemptsService resetLoginAttemptsService;
+    @Mock
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/ResetLoginAttemptsServiceTest.java:28]: add a space above

File: src/test/java/com/zufar/onlinestore/security/api/ResetLoginAttemptsServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    private LoginAttemptRepository loginAttemptRepository;
+    @Mock
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/ResetLoginAttemptsServiceTest.java:30]: add a space above

File: src/test/java/com/zufar/onlinestore/security/api/ResetLoginAttemptsServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    void testLockUserAccount() {
+        when(loginAttemptRepository.findByUserEmail(userEmail))
+                .thenReturn(loginAttemptEntityForId);
+        when(LoginAttemptFactory.createInitialFailedLoggedAttemptEntity(userEmail))
+                .thenReturn(loginAttemptEntity);
+
+        resetLoginAttemptsService.reset(userEmail);
+
+        verify(userAccountLocker, times(1))
+                .unlockUserAccount(userEmail);
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/ResetLoginAttemptsServiceTest.java:59]: one the same line please 
`verify(userAccountLocker, times(1)).unlockUserAccount(userEmail);`

File: src/test/java/com/zufar/onlinestore/security/api/SecurityPrincipalProviderTest.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.mockStatic;
+
+@ExtendWith(MockitoExtension.class)
+class SecurityPrincipalProviderTest {
+    @InjectMocks
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/SecurityPrincipalProviderTest.java:31]: add spaces the same please

File: src/test/java/com/zufar/onlinestore/security/api/SecurityPrincipalProviderTest.java [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+    @AfterAll
+    static void tearDownOnce() {
+        mockedSecurityContextHolder.close();
+    }
+
+    @BeforeEach
+    void setUp() {
+        when(SecurityContextHolder.getContext())
+                .thenReturn(securityContext);
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/SecurityPrincipalProviderTest.java:56]: on the same line please

File: src/test/java/com/zufar/onlinestore/security/api/UserAccountLockerTest.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+@ExtendWith(MockitoExtension.class)
+class UserAccountLockerTest {
+    @InjectMocks
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/UserAccountLockerTest.java:20]: spaces

File: src/test/java/com/zufar/onlinestore/security/api/UserAccountLockerTest.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    private String userEmail = Instancio.create(String.class);
+
+    @Test
+    @DisplayName("Mock test lock user account")
+    void testLockUserAccount() {
+        userAccountLocker.lockUserAccount(userEmail);
+        verify(loginAttemptRepository, times(1))
+                .setUserLockedStatusAndExpiration(eq(userEmail), any());
+        verify(userRepository, times(1))
+                .setAccountLockedStatus(userEmail, false);
```
Comment (Sunagatov) [inline src/test/java/com/zufar/onlinestore/security/api/UserAccountLockerTest.java:35]: on the same line

File: src/test/resources/application-test.yaml [hunk 19]
```diff
@@ -26,4 +26,4 @@ spring:
 jwt:
   header: Authorization
   secret: ${random.uuid}
-  expiration: 1800000
\ No newline at end of file
+  expiration: 1800000
```
Comment (Sunagatov) [inline src/test/resources/application-test.yaml:29]: avoid unnecessary changes here

**Review by Sunagatov — COMMENTED** [review]
Reviewed

File: src/test/java/com/zufar/onlinestore/security/jwt/JwtBlacklistValidatorTest.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
+package com.zufar.onlinestore.security.jwt;
+
+import com.zufar.onlinestore.security.exception.JwtTokenBlacklistedException;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.lang.reflect.Field;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (M437A) [inline src/test/java/com/zufar/onlinestore/security/jwt/JwtBlacklistValidatorTest.java:10]: will open it with a new commit

File: src/test/java/com/zufar/onlinestore/security/jwt/JwtTokenFromAuthHeaderExtractorTest.java [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+    @Mock
+    private HttpServletRequest request;
+
+    private String jwtHttpRequestHeader = HttpHeaders.AUTHORIZATION;
+    private String postfix = "POSTFIX";
+    private String BEARER_PREFIX;
+    private String header;
+
+    @BeforeEach
+    void setUp() throws NoSuchFieldException, IllegalAccessException {
```
Comment (M437A) [inline src/test/java/com/zufar/onlinestore/security/jwt/JwtTokenFromAuthHeaderExtractorTest.java:37]: will remove exceptions in the next commit

File: src/test/java/com/zufar/onlinestore/security/jwt/JwtTokenFromAuthHeaderExtractorTest.java (hunk 1 above)
Comment (M437A) [reply src/test/java/com/zufar/onlinestore/security/jwt/JwtTokenFromAuthHeaderExtractorTest.java]: Fixed

File: .github/workflows/dev-branch-pr-deployment-pipeline.yml [hunk 22]
```diff
@@ -14,10 +14,10 @@ jobs:
...
 
-    - name: Build Docker Image
-      run: |
-        docker login -u ${{secrets.DOCKERHUB_USERNAME}} -p ${{secrets.DOCKERHUB_PASSWORD}}
-        docker build -t zufarexplainedit/test_ci:latest .
-        docker push zufarexplainedit/test_ci:latest
\ No newline at end of file
+      - name: Build Docker Image
+        run: |
+         docker login -u ${{secrets.DOCKERHUB_USERNAME}} -p ${{secrets.DOCKERHUB_PASSWORD}}
```
[the remark this answers is not in the document]
Comment (M437A) [reply .github/workflows/dev-branch-pr-deployment-pipeline.yml:21]: explained

File: src/test/java/com/zufar/onlinestore/security/api/FailedLoginAttemptIncrementorTest.java [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+import static org.mockito.Mockito.mockStatic;
+
+@ExtendWith(MockitoExtension.class)
+class FailedLoginAttemptIncrementorTest {
+    @InjectMocks
+    private FailedLoginAttemptIncrementor failedLoginAttemptIncrementor;
+    @Mock
+    private LoginAttemptRepository loginAttemptRepository;
+    private String userEmail = Instancio.of(String.class)
+            .create();
```
[the remark this answers is not in the document]
Comment (M437A) [reply src/test/java/com/zufar/onlinestore/security/api/FailedLoginAttemptIncrementorTest.java]: Fixed

File: src/test/java/com/zufar/onlinestore/security/api/CustomUserDetailsServiceTest.java [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class CustomUserDetailsServiceTest {
+    @InjectMocks
+    private CustomUserDetailsService customUserDetailsService;
+    @Mock
```
[the remark this answers is not in the document]
Comment (M437A) [reply src/test/java/com/zufar/onlinestore/security/api/CustomUserDetailsServiceTest.java:29]: Fixed

File: src/test/java/com/zufar/onlinestore/security/api/FailedLoginAttemptIncrementorTest.java [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+        mockStatic = mockStatic(LoginAttemptFactory.class);
+    }
+
+    @AfterAll
+    static void tearDownOne() {
+        mockStatic.close();
+    }
+
+    @Test
+    @DisplayName("Incrementing Login Success")
```
[the remark this answers is not in the document]
Comment (M437A) [reply src/test/java/com/zufar/onlinestore/security/api/FailedLoginAttemptIncrementorTest.java]: Fixed
